### PR TITLE
[pbs] Fix extraction (closes #14305)

### LINKED ISSUE #14305: [pbs] Failed to download Nova video
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2017.09.15*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2017.09.15**

```
% youtube-dl --version
2017.09.15
```

### Before submitting an *issue* make sure you have:
- [x] At least skimmed through the [README](https://github.com/rg3/youtube-dl/blob/master/README.md), **most notably** the [FAQ](https://github.com/rg3/youtube-dl#faq) and [BUGS](https://github.com/rg3/youtube-dl#bugs) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?type=Issues) the bugtracker for similar issues including closed ones

### What is the purpose of your *issue*?
- [x] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other

---

### The following sections concretize particular purposed issues, you can erase any section (the contents between triple ---) not applicable to your *issue*

---

### If the purpose of this *issue* is a *bug report*, *site support request* or you are not completely sure provide the full verbose output as follows:

Add the `-v` flag to **your command line** you run youtube-dl with (`youtube-dl -v <your command line>`), copy the **whole** output and insert it here. It should look similar to one below (replace it with **your** log inserted between triple ```):

```
[debug] System config: []
[debug] User config: ['--external-downloader', 'aria2c', '--prefer-ffmpeg', '--ffmpeg-location', '/usr/local/bin/ffmpeg']
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'http://www.pbs.org/wgbh/nova/space/death-dive-to-Saturn.html']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2017.09.15
[debug] Python version 3.6.1 - CYGWIN_NT-10.0-2.9.0-0.318-5-3-x86_64-64bit-WindowsPE
[debug] exe versions: ffmpeg N-85750-ga75ef15, ffprobe N-85750-ga75ef15
[debug] Proxy map: {}
[debug] Using fake IP 3.84.59.216 (US) as X-Forwarded-For.
[pbs] Downloading JSON metadata
[pbs] death-dive-to-Saturn: Downloading webpage
[pbs] Downloading widget/partnerplayer page
WARNING: Unable to download webpage: HTTP Error 404: Not Found
[pbs] Downloading portalplayer page
WARNING: Unable to download webpage: HTTP Error 404: Not Found
ERROR: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/youtube_dl/YoutubeDL.py", line 776, in extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/common.py", line 434, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/pbs.py", line 585, in _real_extract
    self._sort_formats(formats)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/common.py", line 1072, in _sort_formats
    raise ExtractorError('No video formats found')
youtube_dl.utils.ExtractorError: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

---

### Description of your *issue*, suggested solution and other information

Not sure if this is related to #7095 or not, but since #7095 is an "old" issue (opened 2015, last update Jan 2017) with a possible solution #b0db58b, I think a new issue is in order.

I also don't think this is related to #8538, but since youtube-dl totally failed finding any video, probably not.

If you need more info or things to try out, feel free to ask.

### PR DESCRIPTION
**Pull request by gfabiano**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Fix #14305

### CODE REVIEW

**Comment by pepoluan** [comment]
Tested with the exact same URL as in #14305 , still does NOT work. Verbose report follows:

```
[debug] System config: []
[debug] User config: ['--external-downloader', 'aria2c', '--prefer-ffmpeg', '--ffmpeg-location', '/usr/local/bin/ffmpeg']
[debug] Custom config: []
[debug] Command-line args: ['--verbose', 'http://www.pbs.org/wgbh/nova/space/death-dive-to-Saturn.html']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2017.09.24
[debug] Python version 3.6.1 - CYGWIN_NT-10.0-2.9.0-0.318-5-3-x86_64-64bit-WindowsPE
[debug] exe versions: ffmpeg N-87353-g183fd30, ffprobe N-87353-g183fd30
[debug] Proxy map: {}
[debug] Using fake IP 3.48.30.12 (US) as X-Forwarded-For.
[pbs] Downloading JSON metadata
[pbs] death-dive-to-Saturn: Downloading webpage
[pbs] Downloading widget/partnerplayer page
WARNING: Unable to download webpage: HTTP Error 404: Not Found
[pbs] Downloading portalplayer page
WARNING: Unable to download webpage: HTTP Error 404: Not Found
ERROR: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/usr/lib/python3.6/site-packages/youtube_dl/YoutubeDL.py", line 777, in extract_info
    ie_result = ie.extract(url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/common.py", line 434, in extract
    ie_result = self._real_extract(url)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/pbs.py", line 585, in _real_extract
    self._sort_formats(formats)
  File "/usr/lib/python3.6/site-packages/youtube_dl/extractor/common.py", line 1072, in _sort_formats
    raise ExtractorError('No video formats found')
youtube_dl.utils.ExtractorError: No video formats found; please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

**Comment by gfabiano** [comment]
Are you sure my fix does not work?
```
python -m youtube_dl -v http://www.pbs.org/wgbh/nova/space/death-dive-to-Saturn.html
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: [u'-v', u'http://www.pbs.org/wgbh/nova/space/death-dive-to-Saturn.html']
[debug] Encodings: locale cp1252, fs mbcs, out cp850, pref cp1252
[debug] youtube-dl version 2017.09.15
[debug] Git HEAD: cc80dbec9
[debug] Python version 2.7.12 - Windows-10-10.0.15063
[debug] exe versions: ffmpeg 3.3.3, ffprobe 3.3.3
[debug] Proxy map: {}
[debug] Using fake IP 3.192.130.229 (US) as X-Forwarded-For.
[pbs] Downloading JSON metadata
[pbs] death-dive-to-Saturn: Downloading webpage
[pbs] death-dive-to-Saturn: Downloading widget/partnerplayer page
WARNING: Unable to download webpage: HTTP Error 404: Not Found
[pbs] death-dive-to-Saturn: Downloading portalplayer page
WARNING: Unable to download webpage: HTTP Error 404: Not Found
[pbs] death-dive-to-Saturn: Downloading partnerplayer page
[pbs] death-dive-to-Saturn: Downloading hls-1080p-16x9 video url info
[pbs] death-dive-to-Saturn: Downloading m3u8 information
[pbs] death-dive-to-Saturn: Downloading mp4-2500k-16x9 video url info
[pbs] death-dive-to-Saturn: Checking http-2000k video URL
[pbs] death-dive-to-Saturn: http-2000k video URL is invalid, skipping
[pbs] death-dive-to-Saturn: Checking http-6500k video URL
[pbs] death-dive-to-Saturn: http-6500k video URL is invalid, skipping
[pbs] death-dive-to-Saturn: Checking http-4500k video URL
[pbs] death-dive-to-Saturn: http-4500k video URL is invalid, skipping
[pbs] death-dive-to-Saturn: Checking http-3000k video URL
[pbs] death-dive-to-Saturn: Checking http-1100k video URL
[pbs] death-dive-to-Saturn: http-1100k video URL is invalid, skipping
[pbs] death-dive-to-Saturn: Checking http-730k video URL
[pbs] death-dive-to-Saturn: http-730k video URL is invalid, skipping
[debug] Default format spec: bestvideo+bestaudio/best
[debug] Invoking downloader on u'https://ga.video.cdn.pbs.org/videos/nova/3968ea7e-9177-4ba9-aa82-b73c9ef90ee6/2000008498/hd-16x9-mezzanine-1080p/2xpja9ui_nova4412-16x9-1080p-1080p-6500k.m3u8'
```

I download without problems

In your log there isn't this line:
```[pbs] death-dive-to-Saturn: Downloading partnerplayer page```
I assume you probably are not using my fix

**Comment by pepoluan** [comment]
I think I made a mistake somewhere while pulling. After you posted that, I checked my version, and for some reasons it reverted. I'll try another pull and inform you of the results.

**Comment by gfabiano** [comment]
Provided the right support to #14305 and added a test.

**Comment by pepoluan** [comment]
I verify that thix fixes #14305 . LGTM! 😎

**Comment by gfabiano** [comment]
I think I'm ready for a review @dstftw

File: youtube_dl/extractor/pbs.py [hunk 1]
```diff
@@ -415,15 +431,40 @@ def _extract_webpage(self, url):
...
             media_id = self._search_regex(
                 MEDIA_ID_REGEXES, webpage, 'media ID', fatal=False, default=None)
             if media_id:
                 return media_id, presumptive_id, upload_date, description
 
+            media_id = self._search_regex(
+                [r'<(?:section|div)[^>]+data-coveid="([^"]+)"',
+                 r'<input[^>]+id\s*=\s*"pbs_video_id_[0-9]+"\s*value="([^"]+)"[^>]*>',
+                 ],
+                webpage, 'media ID', fatal=False, default=None)
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: `default` implies non `fatal`.

File: youtube_dl/extractor/pbs.py [hunk 2]
```diff
@@ -415,15 +431,40 @@ def _extract_webpage(self, url):
...
+            if media_id:
+                player = self._download_webpage(
+                    urljoin('http://player.pbs.org/partnerplayer/', media_id),
+                    display_id, 'Downloading partnerplayer page', fatal=False)
+                jwSettings = self._parse_json(
+                    self._search_regex(
+                        r'(?s)jwSettings\s*=\s*({.+?});',
+                        player, 'jwSetting data', default='{}'),
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: Breaks when `player` is `False`.

File: youtube_dl/extractor/pbs.py [hunk 3]
```diff
@@ -415,15 +431,40 @@ def _extract_webpage(self, url):
...
                 return media_id, presumptive_id, upload_date, description
 
+            media_id = self._search_regex(
+                [r'<(?:section|div)[^>]+data-coveid="([^"]+)"',
+                 r'<input[^>]+id\s*=\s*"pbs_video_id_[0-9]+"\s*value="([^"]+)"[^>]*>',
+                 ],
+                webpage, 'media ID', fatal=False, default=None)
+            if media_id:
+                player = self._download_webpage(
+                    urljoin('http://player.pbs.org/partnerplayer/', media_id),
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: This should be recursively delegated to pbs extractor instead.

File: youtube_dl/extractor/pbs.py [hunk 4]
```diff
@@ -493,7 +534,7 @@ def extract_redirect_urls(info):
 
         chapters = []
         # Player pages may also serve different qualities
-        for page in ('widget/partnerplayer', 'portalplayer'):
+        for page in ('widget/partnerplayer', 'portalplayer', 'viralplayer'):
```
Comment (dstftw) [inline youtube_dl/extractor/pbs.py]: What's the point of this?

**Pull request closed without merging**